File: jarvix/web_crawler.py
"""
Jarvix - Web Crawler Module v3
Advanced HTML filtering, fact extraction, and knowledge integration.
Strips nav/ads/menus/dictionary-metadata.
Only keeps headings + main prose content.
"""

import re
import time
import logging
from urllib.parse import urlparse, urljoin
from collections import defaultdict
from dataclasses import dataclass, field
from typing import List, Tuple, Optional, Dict, Set
from enum import Enum

try:
    import urllib.request as urllib_req
    from bs4 import BeautifulSoup
except ImportError:
    urllib_req = None
    BeautifulSoup = None

# Configure logging
logger = logging.getLogger(__name__)


class CrawlQuality(Enum):
    """Quality assessment enum for crawl results."""
    A = "A — High-density knowledge (>15% facts)"
    B = "B — Solid factual content (8-15%)"
    C = "C — Moderate factual content (3-8%)"
    D = "D — Low fact density (0-3%)"
    F = "F — No extractable facts"


@dataclass
class PageResult:
    """Result of processing a single page."""
    url: str
    title: str = ""
    word_count: int = 0
    sentence_count: int = 0
    facts_extracted: int = 0
    facts_stored: int = 0
    facts_skipped: int = 0
    top_topics: list = field(default_factory=list)
    error: str = ""
    fetch_time_ms: int = 0


@dataclass
class CrawlReport:
    """Complete report from a crawl operation."""
    seed_url: str
    pages_visited: int = 0
    pages_failed: int = 0
    total_words: int = 0
    total_sentences: int = 0
    total_facts: int = 0
    stored_facts: int = 0
    duplicate_facts: int = 0
    top_topics: list = field(default_factory=list)
    top_facts: list = field(default_factory=list)
    page_results: list = field(default_factory=list)
    knowledge_gain: float = 0.0
    errors: list = field(default_factory=list)
    elapsed_ms: int = 0


class WebCrawler:
    """
    Production-grade web crawler for fact extraction.
    
    Features:
    - Aggressive HTML noise removal (nav, ads, sidebars)
    - Sentence quality filtering
    - SVO triple extraction
    - Confidence-based deduplication
    - Structured reporting
    """

    # Tags always discarded (navigation, ads, UI chrome)
    DISCARD_TAGS = {
        "script", "style", "nav", "header", "footer", "aside",
        "form", "noscript", "iframe", "svg", "button", "input",
        "select", "textarea", "meta", "link", "figure", "figcaption",
        "picture", "video", "audio", "canvas", "map", "object",
        "table", "sup", "sub", "cite",  # dictionary/wiki noise
    }

    # CSS class/id patterns indicating nav/ad/menu content
    NOISE_PATTERNS = re.compile(
        r"(nav|menu|sidebar|banner|ad|ads|advertisement|cookie|"
        r"popup|modal|footer|header|breadcrumb|pagination|"
        r"related|share|social|comment|login|signup|subscribe|"
        r"toc|contents|infobox|hatnote|catlinks|reflist|"
        r"references|external\.links|see\.also)",
        re.I
    )

    # Sentence quality filters - reject junk patterns
    _JUNK_PATTERNS = [
        re.compile(r"^\s*\[[^\]]+\]\s*$"),  # Matches [edit], [1], [citation needed]
        re.compile(r"^[^a-zA-Z]*$"),  # no letters
        re.compile(r"^\s*\d+\s*$"),  # bare numbers
        re.compile(
            r"(click here|read more|learn more|"
            r"sign up|log in|subscribe|cookie|"
            r"privacy policy|terms of)",
            re.I,
        ),
        re.compile(r"^.{1,10}$"),  # too short
        re.compile(r"^.{400,}$"),  # too long
        re.compile(r"\|.*\|"),  # nav pipe separators
        re.compile(
            r"^\s*(home|about|contact|help|search|menu|navigation)\s*$",
            re.I,
        ),
        re.compile(
            r"^(before|after|used|when|consonant sound|vowel sound)\s+",
            re.I,
        ),
        re.compile(
            r"\b(IPA|pronunciation|phonetic|syllable|hyphenation)\b",
            re.I,
        ),
    ]

    # Patterns for Subject-Verb-Object (SVO) extraction
    _RELATION_PATTERNS = [
        (re.compile(r"\b([A-Z][a-zA-Z0-9_\s]{2,40})\s+(is a|is an|are|was a|was an|were)\s+([a-zA-Z0-9_\s]{2,60})", re.I), "is_a"),
        (re.compile(r"\b([A-Z][a-zA-Z0-9_\s]{2,40})\s+(has|have|contains|includes)\s+([a-zA-Z0-9_\s]{2,60})", re.I), "has"),
        (re.compile(r"\b([A-Z][a-zA-Z0-9_\s]{2,40})\s+(causes|leads to|results in)\s+([a-zA-Z0-9_\s]{2,60})", re.I), "causes"),
        (re.compile(r"\b([A-Z][a-zA-Z0-9_\s]{2,40})\s+(can|is able to)\s+([a-zA-Z0-9_\s]{2,60})", re.I), "can"),
        (re.compile(r"\b([A-Z][a-zA-Z0-9_\s]{2,40})\s+(is located in|is part of)\s+([a-zA-Z0-9_\s]{2,60})", re.I), "part_of"),
    ]

    # Configuration thresholds
    MIN_PHRASE_LENGTH = 2
    MAX_SUBJECT_LENGTH = 60
    MAX_OBJECT_LENGTH = 120
    MAX_OBJECT_WORDS = 12
    MIN_REAL_WORDS_IN_SENTENCE = 3
    DUPLICATE_CONFIDENCE_THRESHOLD = 0.5
    BASE_TRIPLE_CONFIDENCE = 0.60
    REQUEST_TIMEOUT_SECONDS = 8
    DEFAULT_USER_AGENT = "Jarvix-Crawler/3.0"

    def __init__( 
        self,
        agent,
        max_depth: int = 1,
        max_pages: int = 10,
        timeout_s: int = 8,
        same_domain_only: bool = True,
        request_delay_s: float = 0.5,
    ):
        """Initialize the web crawler."""
        self.agent = agent
        self.max_depth = max_depth
        self.max_pages = max_pages
        self.timeout_s = timeout_s
        self.same_domain_only = same_domain_only
        self.request_delay_s = request_delay_s
        self._visited: Set[str] = set()
        self._last_request_time = 0.0

    def _rate_limit(self):
        """Politeness delay between outbound HTTP requests."""
        elapsed = time.time() - self._last_request_time
        if elapsed < self.request_delay_s:
            time.sleep(self.request_delay_s - elapsed)
        self._last_request_time = time.time()

    def fetch_url(self, url: str) -> Tuple[Optional[str], str]:
        """Fetches URL contents safely with timeouts."""
        if not urllib_req:
            return None, "urllib not available"

        self._rate_limit()
        req = urllib_req.Request(
            url,
            headers={"User-Agent": self.DEFAULT_USER_AGENT}
        )
        try:
            with urllib_req.urlopen(req, timeout=self.timeout_s) as response:
                if not response or not hasattr(response, "headers") or response.headers is None:
                    return None, "Invalid response headers"
                
                content_type = response.headers.get("Content-Type", "") or ""
                if "text/html" not in content_type and "text/plain" not in content_type:
                    return None, f"Unsupported Content-Type: {content_type}"
                
                body = response.read()
                if body is None:
                    return None, "Empty body received"
                
                html_data = body.decode("utf-8", errors="ignore")
                return html_data, ""
        except Exception as e:
            return None, str(e)

    def clean_html(self, html_content: Optional[str]) -> Tuple[str, str, List[str]]:
        """Strips noise tags, headers, nav, ads and returns clean text + inner links."""
        if not BeautifulSoup or not html_content:
            return "", "", []

        soup = BeautifulSoup(html_content, "html.parser")
        if not soup:
            return "", "", []

        title = ""
        if soup.title and soup.title.string:
            title = soup.title.string.strip()

        # Remove discarded noise tags safely
        for tag in soup.find_all(self.DISCARD_TAGS):
            if tag:
                tag.decompose()

        # Remove elements matching noise patterns
        for element in soup.find_all(True):
            if not element:
                continue
            
            # Safe attribute extractions guarded against None
            elem_id = element.get("id") or ""
            classes = element.get("class") or []
            elem_class = " ".join(classes) if isinstance(classes, list) else str(classes)

            if self.NOISE_PATTERNS.search(elem_id) or self.NOISE_PATTERNS.search(elem_class):
                element.decompose()

        # Extract links safely before dropping link text
        links = []
        for a_tag in soup.find_all("a", href=True):
            if a_tag and a_tag.get("href"):
                href = a_tag.get("href")
                if href:
                    links.append(href)

        text = soup.get_text(separator=" ") or ""
        # Normalize whitespace
        clean_text = re.sub(r"\s+", " ", text).strip()
        return title, clean_text, links

    def is_valid_sentence(self, sentence: Optional[str]) -> bool:
        """Filters out junk lines, navigation links, and unreadable text."""
        if not sentence:
            return False

        sentence_str = sentence.strip()
        if not sentence_str:
            return False

        for pattern in self._JUNK_PATTERNS:
            if pattern.search(sentence_str):
                return False

        words = [w for w in sentence_str.split() if w.isalpha()]
        if len(words) < self.MIN_REAL_WORDS_IN_SENTENCE:
            return False

        return True

    def extract_facts_from_text(self, text: Optional[str]) -> List[Tuple[str, str, str]]:
        """Extracts SVO triples from clean text."""
        if not text:
            return []

        sentences = re.split(r"(?<=[.!?])\s+", text)
        facts = []

        for sent in sentences:
            if not self.is_valid_sentence(sent):
                continue

            for pattern, relation in self._RELATION_PATTERNS:
                match = pattern.search(sent)
                if match:
                    groups = match.groups()
                    if groups and len(groups) >= 3:
                        subj = (groups[0] or "").strip()
                        obj = (groups[2] or "").strip()
                        if (
                            subj and obj
                            and len(subj) <= self.MAX_SUBJECT_LENGTH
                            and len(obj) <= self.MAX_OBJECT_LENGTH
                            and len(obj.split()) <= self.MAX_OBJECT_WORDS
                        ):
                            facts.append((subj, relation, obj))
                            break

        return facts

    def crawl(self, seed_url: str) -> CrawlReport:
        """
        Executes web crawl starting from seed_url.
        Processes pages, cleans HTML, extracts facts, and integrates into memory.
        """
        start_time = time.time()
        report = CrawlReport(seed_url=seed_url)
        queue = [(seed_url, 0)]
        seed_domain = urlparse(seed_url).netloc or ""

        while queue and report.pages_visited < self.max_pages:
            url, depth = queue.pop(0)

            # Normalize URL
            url_clean = url.split("#")[0].rstrip("/")
            if url_clean in self._visited:
                continue

            self._visited.add(url_clean)

            page_start = time.time()
            html, err = self.fetch_url(url_clean)
            fetch_time_ms = int((time.time() - page_start) * 1000)

            if err:
                report.pages_failed += 1
                report.errors.append(f"{url_clean}: {err}")
                report.page_results.append(PageResult(url=url_clean, error=err, fetch_time_ms=fetch_time_ms))
                continue

            title, clean_text, links = self.clean_html(html)
            words = (clean_text or "").split()
            word_count = len(words)
            sentences = [s for s in re.split(r"(?<=[.!?])\s+", clean_text) if self.is_valid_sentence(s)]
            
            facts = self.extract_facts_from_text(clean_text)
            
            stored_count = 0
            skipped_count = 0

            # Integrate into Agent Memory safely
            for s, r, o in facts:
                mem = getattr(self.agent, "semantic_memory", None)
                if mem and hasattr(mem, "add_edge"):
                    edge = mem.add_edge(
                        subject=s,
                        relation=r,
                        object_=o,
                        confidence=self.BASE_TRIPLE_CONFIDENCE,
                        source=url_clean
                    )
                    if edge:
                        stored_count += 1
                    else:
                        skipped_count += 1

            report.pages_visited += 1
            report.total_words += word_count
            report.total_sentences += len(sentences)
            report.total_facts += len(facts)
            report.stored_facts += stored_count
            report.duplicate_facts += skipped_count

            report.page_results.append(PageResult(
                url=url_clean,
                title=title,
                word_count=word_count,
                sentence_count=len(sentences),
                facts_extracted=len(facts),
                facts_stored=stored_count,
                facts_skipped=skipped_count,
                fetch_time_ms=fetch_time_ms
            ))

            # Queue child links if within max depth
            if depth < self.max_depth:
                for link in links:
                    if not link:
                        continue
                    full_link = urljoin(url_clean, link).split("#")[0].rstrip("/")
                    link_domain = urlparse(full_link).netloc or ""

                    if self.same_domain_only and link_domain != seed_domain:
                        continue

                    if full_link not in self._visited:
                        queue.append((full_link, depth + 1))

        report.elapsed_ms = int((time.time() - start_time) * 1000)
        return report

File: server/jarvix/web-crawler.js
import * as cheerio from 'cheerio';
import { parse } from './nlp-parser.js';

const DISCARD_TAGS = ['nav', 'script', 'style', 'footer', 'aside', 'form', 'iframe', 'noscript', 'button', 'select', 'option', 'label', 'figure', 'figcaption', 'img', 'svg', 'video', 'audio', 'source', 'track', 'map', 'area'];

const NOISE_PATTERNS = /nav|ad-banner|cookie-notice|infobox|reflist|metadata|sidebar|footer-box|header-box|menu-bar|banner-ad|popup|social-share|share-buttons|comment-section|related-articles|navigation-bar|edit-section|reference-list|citation-needed/i;

const JUNK_PATTERNS = [
  /^(click|buy|download|sign|subscribe|share|follow|like|log|register|join|learn more|read more|see more|view more)/i,
  /^\d+$/,
  /^[^a-z]*$/i,
  /^.{1,15}$/,
  /^.{200,}$/,
  /\b(ipa|pronunciation|phonetic)\b/i,
  /^\[[^\]]*\]$/,
  /^(categories|references|external links|see also|contents|further reading)$/,
  /^citation needed/i,
  /^\d{4}-\d{2}-\d{2}/,
];

const REL_PATTERNS = [
  [/^(.+?)\s+(?:is|are)\s+(?:a|an)\s+(.+)$/i, 'is_a'],
  [/^(.+?)\s+(?:is|are)\s+(.+)$/i, 'has_property'],
  [/^(.+?)\s+has\s+(.+)$/i, 'has'],
  [/^(.+?)\s+have\s+(.+)$/i, 'has'],
  [/^(.+?)\s+can\s+(.+)$/i, 'can'],
  [/^(.+?)\s+causes\s+(.+)$/i, 'causes'],
  [/^(.+?)\s+contains\s+(.+)$/i, 'has'],
  [/^(.+?)\s+includes\s+(.+)$/i, 'has'],
  [/^(.+?)\s+requires\s+(.+)$/i, 'has'],
  [/^(.+?)\s+consists of\s+(.+)$/i, 'part_of'],
  [/^(.+?)\s+is part of\s+(.+)$/i, 'part_of'],
  [/^(.+?)\s+refers to\s+(.+)$/i, 'related_to'],
  [/^(.+?)\s+means\s+(.+)$/i, 'related_to'],
  [/^(.+?)\s+uses\s+(.+)$/i, 'does'],
  [/^(.+?)\s+produces\s+(.+)$/i, 'causes'],
  [/^(.+?)\s+belongs to\s+(.+)$/i, 'part_of'],
  [/^(.+?)\s+was\s+(?:a|an)\s+(.+)$/i, 'is_a'],
];

const MIN_PHRASE_LENGTH = 2;
const MAX_SUBJECT_LENGTH = 60;
const MAX_OBJECT_LENGTH = 120;
const MAX_OBJECT_WORDS = 12;
const DUPLICATE_CONFIDENCE_THRESHOLD = 0.5;
const BASE_TRIPLE_CONFIDENCE = 0.60;
const DEFAULT_USER_AGENT = 'Jarvix-Crawler/3.0';

function cleanPhrase(s) {
  return s.toLowerCase().replace(/[^a-z0-9 '\-]/g, ' ').replace(/\s+/g, ' ').trim().replace(/^(a|an|the)\s+/, '');
}

function splitSentences(text) {
  return text.split(/[.!?]\s+/).map(s => s.trim()).filter(Boolean);
}

function isJunk(sentence) {
  for (const pat of JUNK_PATTERNS) if (pat.test(sentence)) return true;
  const words = sentence.split(/\s+/).filter(w => w.length > 2 && /^[a-z]/i.test(w));
  if (words.length < 3) return true;
  const firstWord = sentence.split(/\s+/)[0];
  if (firstWord.length <= 2 && firstWord.toLowerCase() !== 'i' && firstWord.toLowerCase() !== 'a') return true;
  return false;
}

function extractTriple(sentence) {
  for (const [pattern, rel] of REL_PATTERNS) {
    const m = sentence.match(pattern);
    if (m) {
      let subj = cleanPhrase(m[1]);
      let obj = cleanPhrase(m[2]);
      if (subj.length < MIN_PHRASE_LENGTH || subj.length > MAX_SUBJECT_LENGTH) continue;
      if (obj.length < MIN_PHRASE_LENGTH || obj.length > MAX_OBJECT_LENGTH) continue;
      if (obj.split(/\s+/).length > MAX_OBJECT_WORDS) continue;
      if (PRONOUNS.has(subj.split(/\s+/)[0])) continue;
      if (!subj || !obj) continue;
      return { subject: subj, relation: rel, object: obj };
    }
  }
  return null;
}

const PRONOUNS = new Set(['he', 'she', 'it', 'they', 'this', 'that', 'these', 'those', 'which', 'who', 'what', 'there', 'here']);

function computeQuality(report) {
  const ratio = report.totalSentences > 0 ? report.storedFacts / report.totalSentences : 0;
  if (report.storedFacts === 0) return 'F';
  if (ratio > 0.15) return 'A';
  if (ratio > 0.08) return 'B';
  if (ratio > 0.03) return 'C';
  return 'D';
}

export class WebCrawler {
  constructor(agent, { maxDepth = 1, maxPages = 10, timeoutS = 8, sameDomainOnly = true, requestDelayS = 0.5 } = {}) {
    this.agent = agent;
    this.maxDepth = maxDepth;
    this.maxPages = maxPages;
    this.timeoutS = timeoutS;
    this.sameDomainOnly = sameDomainOnly;
    this.requestDelayS = requestDelayS;
    this._visited = new Set();
    this._lastRequestTime = 0;
  }

  async _fetch(url) {
    const now = Date.now();
    const elapsed = now - this._lastRequestTime;
    const delay = this.requestDelayS * 1000;
    if (elapsed < delay) await new Promise(r => setTimeout(r, delay - elapsed));
    this._lastRequestTime = Date.now();

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutS * 1000);
    try {
      const res = await fetch(url, {
        headers: { 'User-Agent': DEFAULT_USER_AGENT },
        signal: controller.signal,
      });
      const html = await res.text();
      return { html, status: res.status };
    } catch (e) {
      return { html: null, status: 0, error: e.message };
    } finally {
      clearTimeout(timer);
    }
  }

  _parseHtml(html) {
    const $ = cheerio.load(html);
    for (const tag of DISCARD_TAGS) $(tag).remove();
    $('[class]').each((_, el) => {
      const cls = $(el).attr('class') || '';
      const id = $(el).attr('id') || '';
      if (NOISE_PATTERNS.test(cls) || NOISE_PATTERNS.test(id)) $(el).remove();
    });
    const title = $('title').text().trim() || $('h1').first().text().trim() || '';
    const content = $('main').text() || $('article').text() || $('body').text() || '';
    const paragraphs = [];
    $('p, h1, h2, h3, h4, h5, h6, li').each((_, el) => {
      const text = $(el).text().trim();
      if (text) paragraphs.push(text);
    });
    const text = paragraphs.length ? paragraphs.join('\n') : content;
    return { title, text };
  }

  _extractLinks(html, baseUrl, seedDomain) {
    const $ = cheerio.load(html);
    const links = [];
    $('a[href]').each((_, el) => {
      const href = $(el).attr('href');
      if (!href) return;
      try {
        const fullUrl = new URL(href, baseUrl).href;
        if (!fullUrl.startsWith('http://') && !fullUrl.startsWith('https://')) return;
        if (this.sameDomainOnly) {
          const linkDomain = new URL(fullUrl).hostname;
          if (linkDomain !== seedDomain) return;
        }
        if (!links.includes(fullUrl)) links.push(fullUrl);
      } catch {}
    });
    return links.slice(0, 5);
  }

  _processPage(url, seedDomain) {
    const pageStart = Date.now();
    return this._fetch(url).then(({ html, status, error }) => {
      if (error || !html || status >= 400) {
        return {
          url, title: '', wordCount: 0, sentenceCount: 0,
          factsExtracted: 0, factsStored: 0, factsSkipped: 0,
          topTopics: [], error: error || `HTTP ${status}`, fetchTimeMs: Date.now() - pageStart,
        };
      }
      const { title, text } = this._parseHtml(html);
      const sentences = splitSentences(text);
      const cleanSentences = sentences.filter(s => !isJunk(s));
      const wordCount = text.split(/\s+/).filter(Boolean).length;
      let stored = 0, skipped = 0;
      const topics = new Set();

      for (const sentence of cleanSentences.slice(0, 200)) {
        const triple = extractTriple(sentence);
        if (!triple) continue;
        const existingConf = this.agent.semanticMemory.edgeConfidence(triple.subject, triple.relation, triple.object);
        if (existingConf > DUPLICATE_CONFIDENCE_THRESHOLD) { skipped++; continue; }

        this.agent.semanticMemory.addEdge(triple.subject, triple.relation, triple.object, BASE_TRIPLE_CONFIDENCE, false, 'web');
        this.agent.brain.graph.addEdge(triple.subject, triple.relation, triple.object, BASE_TRIPLE_CONFIDENCE, false, 'web');
        this.agent.memory.addFact(triple.subject, `${triple.relation.replace(/_/g, ' ')} ${triple.object}`, BASE_TRIPLE_CONFIDENCE);
        stored++;
        topics.add(triple.subject);
      }

      let links = [];
      if (this.maxDepth > 0) {
        links = this._extractLinks(html, url, seedDomain);
      }

      return {
        url, title, wordCount, sentenceCount: cleanSentences.length,
        factsExtracted: stored + skipped, factsStored: stored, factsSkipped: skipped,
        topTopics: [...topics].slice(0, 5), error: null, fetchTimeMs: Date.now() - pageStart,
        _links: links,
      };
    });
  }

  async crawl(seedUrl) {
    const start = Date.now();
    const factsBefore = Object.values(this.agent.memory.facts).flatMap(t => Object.keys(t)).length;
    let seedDomain;
    try { seedDomain = new URL(seedUrl).hostname; } catch { seedDomain = ''; }

    const report = {
      seedUrl, pagesVisited: 0, pagesFailed: 0, totalWords: 0,
      totalSentences: 0, totalFacts: 0, storedFacts: 0, duplicateFacts: 0,
      topTopics: [], topFacts: [], pageResults: [], knowledgeGain: 0,
      errors: [], elapsedMs: 0,
    };

    const queue = [{ url: seedUrl, depth: 0 }];

    while (queue.length > 0 && report.pagesVisited + report.pagesFailed < this.maxPages) {
      const { url, depth } = queue.shift();
      if (this._visited.has(url)) continue;
      this._visited.add(url);

      const result = await this._processPage(url, seedDomain);
      if (result.error) {
        report.pagesFailed++;
        report.errors.push(`${url}: ${result.error}`);
        report.pageResults.push(result);
        continue;
      }

      report.pagesVisited++;
      report.totalWords += result.wordCount;
      report.totalSentences += result.sentenceCount;
      report.totalFacts += result.factsExtracted;
      report.storedFacts += result.factsStored;
      report.duplicateFacts += result.factsSkipped;
      report.pageResults.push(result);

      if (result._links && depth < this.maxDepth) {
        for (const link of result._links) {
          if (!this._visited.has(link)) queue.push({ url: link, depth: depth + 1 });
        }
      }
    }

    report.elapsedMs = Date.now() - start;
    const factsAfter = Object.values(this.agent.memory.facts).flatMap(t => Object.keys(t)).length;
    report.knowledgeGain = Math.round((factsAfter - factsBefore) / Math.max(factsBefore, 1) * 100);

    const topicCounts = {};
    for (const pr of report.pageResults) {
      for (const t of pr.topTopics || []) topicCounts[t] = (topicCounts[t] || 0) + 1;
    }
    report.topTopics = Object.entries(topicCounts).sort((a, b) => b[1] - a[1]).slice(0, 10).map(([t]) => t);

    const recentEdges = this.agent.semanticMemory.edges.filter(e => e.source === 'web').slice(-20);
    report.topFacts = recentEdges.slice(0, 12).map(e => ({
      subject: e.subject, relation: e.relation, object: e.object_, confidence: e.confidence,
    }));

    return report;
  }

  buildEvaluation(report) {
    return {
      summary: {
        seed_url: report.seedUrl,
        pages_visited: report.pagesVisited,
        pages_failed: report.pagesFailed,
        total_words: report.totalWords,
        total_sentences: report.totalSentences,
        facts_extracted: report.totalFacts,
        facts_stored: report.storedFacts,
        duplicates: report.duplicateFacts,
        knowledge_gain: report.knowledgeGain,
        elapsed_ms: report.elapsedMs,
        quality_score: computeQuality(report),
      },
      top_topics: report.topTopics,
      top_facts: report.topFacts,
      page_results: report.pageResults.map(p => ({
        url: p.url, title: p.title, words: p.wordCount, sentences: p.sentenceCount,
        stored: p.factsStored, skipped: p.factsSkipped, top_topics: p.topTopics || [],
        error: p.error, ms: p.fetchTimeMs,
      })),
      errors: report.errors,
      inference_note: report.storedFacts > 0
        ? `Inferred ${this.agent.brain.graph.stats().inferred_edges} additional facts through transitive reasoning.`
        : '',
    };
  }
}
